pkg/virt-api/rest/streamer.go

```diff
@@ -0,0 +1,162 @@
+package rest
+
+import (
+	"context"
+	"io"
+	"net"
+	"time"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/gorilla/websocket"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+)
+
+type vmiFetcher func(namespace, name string) (*v1.VirtualMachineInstance, *errors.StatusError)
+type dialer func(vmi *v1.VirtualMachineInstance) (net.Conn, *errors.StatusError)
+type validator func(vmi *v1.VirtualMachineInstance) *errors.StatusError
+type streamFunc func(clientConn *websocket.Conn, serverConn net.Conn, result chan<- streamFuncResult)
+type streamFuncResult error
+
+type Streamer struct {
+	fetchVMI        vmiFetcher
+	validateVMI     validator
+	dial            dialer
+	keepAliveClient func(ctx context.Context, conn *websocket.Conn, cancel func())
```

Comment:
- [ ] Add comment explaining that keepAliveClient only works for streaming that uses the websocket and not its underlying connection.